### PR TITLE
[15] Stage 3: Supply registration identifier

### PR DESCRIPTION
**Pull request by edavey**
[Trello 15](https://trello.com/c/BvEYSrto/15-stage-3-supply-spacecraft-registration-identifier)

## Stage 3: Pilot supplies spacecraft registration identifier

The registration identifier needs to be in the `XXX111X` format.

See individual commit messages for details.

stage_3_supply_registration_information

### CODE REVIEW

**Comment by edavey** [comment]
[Stage 4: Supply personal details](https://trello.com/c/hMAy7QYS/38-stage-4-supply-personal-details)

File: app/forms/spacecraft_registration_identifier_form.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class SpacecraftRegistrationIdentifierForm
+  include ActiveModel::Model
+
+  attr_accessor :sri
+
+  def initialize(sri:)
+    @sri = sri
```
Comment (Gweaton) [inline app/forms/spacecraft_registration_identifier_form.rb:7]: What are your thoughts on using the full `spacecraft_registration_identifier` here? I'm a fan of avoiding acronyms where possible, but if it's common terminology throughout the service, maybe it's not a problem?

File: app/views/stages/registration_identifier_stage/show.html.erb [hunk 2]
```diff
@@ -3,6 +3,17 @@
     <div class="row">
       <article class="col">
         <h1 class="govuk-heading-xl">Your registration identifier</h1>
+
+        <p class="govuk-body">We need to record your Spacecraft Registration Identifier (SRI).</p>
```
Comment (Gweaton) [inline app/views/stages/registration_identifier_stage/show.html.erb:7]: Ah, we do refer to an `SRI` here, so maybe disregard my previous comment.

File: spec/features/pilot/stage_3_supply_registration_identifier_spec.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    when_i_provide_spacecraft_registration_identifier
+    and_i_proceed_to_the_next_stage
+    then_i_should_find_myself_at_the_personal_details_stage
+  end
+
+  def given_i_am_on_the_registration_identifier_stage
+    visit("stages/registration-identifier")
+  end
+
+  def when_i_fail_to_provide_an_valid_reg
```
Comment (Gweaton) [inline spec/features/pilot/stage_3_supply_registration_identifier_spec.rb:47]: `...a_valid_reg`?

File: spec/features/pilot/stage_3_supply_registration_identifier_spec.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+# Feature: Stage 3: Supply "Spacecraft Registration Identifier"
+#   So that my "reg" can be checked by "the authorities"
+#   As a pilot
+#   I want to supply a validated "Spacecraft Registration Identifier"
+
+RSpec.feature "Stage 3: Supply registration identifier" do
+  # Scenario: Must supply Spacecraft Registration Identifier
+  #   Given I am on the 'registration identifier' stage
```
Comment (Gweaton) [inline spec/features/pilot/stage_3_supply_registration_identifier_spec.rb:10]: Is using comments for Cucumber specs a common pattern? I've seen it on a few projects, but would have thought that the verbosity of the method names below would hopefully remove the need for additional comments. More of a genuine question than something to change!

**Review by Gweaton — APPROVED** [review]
The approach looks good to me - just a couple of questions/comments but no blockers.

File: spec/features/pilot/stage_3_supply_registration_identifier_spec.rb (hunk 3 above)
Comment (edavey) [reply spec/features/pilot/stage_3_supply_registration_identifier_spec.rb:47]: thks, well spotted! On reflection I actually think "valid identifier" is clearer so am changing to that, as well as fixing that typo...

File: app/forms/spacecraft_registration_identifier_form.rb (hunk 1 above)
Comment (edavey) [reply app/forms/spacecraft_registration_identifier_form.rb:7]: good point, I also hate acronyms -- a huge source of obscurity...

Reviewing our See [Standard Terminology (ADR 12)](https://github.com/dxw/dfsseta-apply-for-landing-ruby/blob/main/doc/architecture/decisions/0012-standard-terminology.md?plain=1#L38-L39) I see that "Registration ID" is the right shortening, so I'm going to apply that...

File: app/views/stages/registration_identifier_stage/show.html.erb (hunk 2 above)
Comment (edavey) [reply app/views/stages/registration_identifier_stage/show.html.erb:7]: I'm updating this heading to "Your registration ID" (see above)

File: spec/features/pilot/stage_3_supply_registration_identifier_spec.rb (hunk 4 above)
Comment (edavey) [reply spec/features/pilot/stage_3_supply_registration_identifier_spec.rb:10]: The technique which I try to use and to encourage is to:

1. Write a BDD-style acceptance/feature/end-to-end test in the user story /  Trello / Jira ticket
2. Copy this word for word into the high-level testing framework of use, e.g. Cucumber, Rspec-Capybara, Cypress

In the case of Cucumber, the plain English spec becomes  the executable test file. In other frameworks like Rspec/Capybara (in use here) we need to turn our lines of `given/when/then/and` statements into method calls and write the matching methods.

I think you're right that once the method-calls and methods are implemented, the original comments in plain English become largely redundant and a source of maintenance error. i.e. they can become out of sync with the methods themselves. 

You don't have this issue if you use Cucumber, which I think is the best approach. I believe we ought to make Cucumber the default choice of high-level testing framework.

Focussing on the high level benefits and functionality of features is critical to delivering software which is user-centric. Cucumber is in my view the best tool to drive this practice.

Comment (edavey) [reply spec/features/pilot/stage_3_supply_registration_identifier_spec.rb]: I've removed the comments from the scenarios and left the "feature" comment:

```
# Feature: Stage 3: Supply "Spacecraft Registration Identifier"
#   So that my "reg" can be checked by "the authorities"
#   As a pilot
#   I want to supply a validated "Spacecraft Registration Identifier"
```
as that does have enduring value to state the purpose of the feature.
